File: Java/src/TaskExecutor.java
import java.io.File;
import java.util.LinkedList;
import java.util.Queue;

public class TaskExecutor {
    int N;
    int E;
    int T;

    Queue<Tarefa> Tarefas;
    Queue<Resultado> Resultados;

    Arquivo arquivo;

    public TaskExecutor(int N, int E, int T, File caminho){
        this.N = N;
        this.E = E;
        this.T = T;

        Tarefas = new LinkedList<Tarefa>();
        Resultados = new LinkedList<Resultado>();

        arquivo = new Arquivo(caminho);
    }

    public void iniciar(){
        alimentarTarefas();
        long tempoInicioTeste = System.currentTimeMillis();
        Executor executor = new Executor(this);
        executor.start();
        
        iniciarTrabalhadores(executor);

        long tempoFinalTeste = System.currentTimeMillis() - tempoInicioTeste;
        System.out.println(tempoFinalTeste + " ms");
        Main.somarTempoTestes(tempoFinalTeste);

    }

    public void alimentarTarefas(){
        for(long i = 0; i < Math.pow(10, N); i++){
            double custo = Math.random() * 0.01;
            int tipo = Math.random() <= (double)E/100 ? 0 : 1;
            int valor = (int) Math.floor(Math.random() * 11);
            Tarefas.add(new Tarefa(i, custo, tipo, valor));
        }
    }

    public void iniciarTrabalhadores(Executor executor){
        Trabalhador[] trabalhadores = new Trabalhador[T];

        for (int i = 0; i < T; i++) {
			trabalhadores[i] = new Trabalhador(executor, arquivo);
		}
		for (int i = 0; i < T; i++) {
			trabalhadores[i].start();
            trabalhadores[i].setPriority(10);
		}
        
		try {
			for (int i = 0; i < T; i++) {
				trabalhadores[i].join();
			}
		} catch (InterruptedException e) {
			e.printStackTrace();
		}
    }
}

File: GO/src/classes/TaskExecutor.go
package classes

import (
	"fmt"
	"math"
	"math/rand"
	"sync"
	"time"
	"strconv"
	"os"
)

type TaskExecutor struct {
	N          int
	E          int
	T          int
	Tarefas    []*Tarefa
	Resultados []*Resultado
	Arquivo    *Arquivo
}

func CriarTaskExecutor(n int, e int, t int, resultadosValores *os.File) TaskExecutor{
	return TaskExecutor{
		N: n,
		E: e,
		T: t,
		Arquivo: CriarArquivo(resultadosValores),
	}
}

func (te *TaskExecutor) Iniciar() int64{
	alimentarTarefasInicio := time.Now();
	te.AlimentarTarefas()
	fmt.Println("Alimentar Tarefas = " + strconv.FormatInt(time.Since(alimentarTarefasInicio).Milliseconds(), 10));
	canalBufferizado := make(chan *Tarefa, (te.T + 50))

	tempoInicioTeste := time.Now()
	executor := CriarExecutor(te, canalBufferizado)
	go executor.Run()

	te.IniciarTrabalhadores(executor, canalBufferizado)

	tempoFinalTeste := time.Since(tempoInicioTeste).Milliseconds()

	
	
	return tempoFinalTeste
}

func (te *TaskExecutor) AlimentarTarefas() {
	potenciaDe10 := int64(math.Pow(float64(10), float64(te.N)))

	for i := int64(0); i < potenciaDe10; i++ {
		custo := rand.Float64() * 0.01
		tipo := 0
		if rand.Float64() <= float64(te.E)/100 {
			tipo = 1
		}
		valor := rand.Intn(11)
		te.Tarefas = append(te.Tarefas, CriarTarefa(i, custo, tipo, valor))
	}
}

func (te *TaskExecutor) IniciarTrabalhadores(executor *Executor, canalBufferizado <-chan *Tarefa) {
	var wg sync.WaitGroup

	inicioTrabalhadores := time.Now();
	for i := 0; i < te.T; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			trabalhador := CriarTrabalhador(te.Arquivo, executor, canalBufferizado)
			trabalhador.Run()
		}()
	}
	fmt.Println("iniciar trabalhadores = " + strconv.FormatInt(time.Since(inicioTrabalhadores).Milliseconds(), 10));
	wg.Wait()
}